pkg/network/p2p/manager.go

```diff
@@ -48,125 +67,142 @@ func WithNoDefaultTimeout() ConnectPeerOption {
 }
 
-// UnregisterProtocol unregisters a protocol.
-func (m *Manager) UnregisterProtocol(protocolID string) {
-	m.registeredProtocolsMutex.Lock()
-	defer m.registeredProtocolsMutex.Unlock()
+// UnregisterProtocol unregisters the handlers for the protocol.
+func (m *Manager) UnregisterProtocol() {
+	m.protocolHandlerMutex.Lock()
+	defer m.protocolHandlerMutex.Unlock()
 
 	m.libp2pHost.RemoveStreamHandler(protocol.ID(protocolID))
-	delete(m.registeredProtocols, protocol.ID(protocolID))
+	m.protocolHandler = nil
 }
 
-// P2PHost returns the lib-p2p host.
-func (m *Manager) P2PHost() host.Host {
-	return m.libp2pHost
-}
+// DialPeer connects to a peer.
+func (m *Manager) DialPeer(ctx context.Context, peer *network.Peer, opts ...ConnectPeerOption) error {
+	m.protocolHandlerMutex.RLock()
+	defer m.protocolHandlerMutex.RUnlock()
 
-// LocalPeer return the local peer.
-func (m *Manager) LocalPeer() *peer.Local {
-	return m.local
-}
+	if m.neighborExists(peer.ID) {
+		return ierrors.Wrapf(ErrDuplicateNeighbor, "peer %s already exists", peer.ID)
+	}
 
-// AddOutbound tries to add a neighbor by connecting to that peer.
-func (m *Manager) AddOutbound(ctx context.Context, p *peer.Peer, group NeighborsGroup,
-	connectOpts ...ConnectPeerOption,
-) error {
-	m.log.Debugw("adding outbound neighbor", "peer", p.ID())
-	return m.addNeighbor(ctx, p, group, m.dialPeer, connectOpts)
-}
+	conf := buildConnectPeerConfig(opts)
+
+	// Adds the peer's multiaddresses to the peerstore, so that they can be used for dialing.
+	m.libp2pHost.Peerstore().AddAddrs(peer.ID, peer.PeerAddresses, peerstore.ConnectedAddrTTL)
+	cancelCtx := ctx
+	if conf.useDefaultTimeout {
+		var cancel context.CancelFunc
+		cancelCtx, cancel = context.WithTimeout(ctx, defaultConnectionTimeout)
+		defer cancel()
+	}
+
+	stream, err := m.P2PHost().NewStream(cancelCtx, peer.ID, protocolID)
+	if err != nil {
+		return ierrors.Wrapf(err, "dial %s / %s failed to open stream for proto %s", peer.PeerAddresses, peer.ID, protocolID)
+	}
+
+	if m.protocolHandler == nil {
+		return ierrors.New("no protocol handler registered")
+	}
+
+	ps := NewPacketsStream(stream, m.protocolHandler.PacketFactory)
+	if err := ps.sendNegotiation(); err != nil {
+		m.closeStream(stream)
+
+		return ierrors.Wrapf(err, "dial %s / %s failed to send negotiation for proto %s", peer.PeerAddresses, peer.ID, protocolID)
+	}
+
+	m.log.Debugw("outgoing stream negotiated",
+		"id", peer.ID,
+		"addr", ps.Conn().RemoteMultiaddr(),
+		"proto", protocolID,
+	)
 
-// AddInbound tries to add a neighbor by accepting an incoming connection from that peer.
-func (m *Manager) AddInbound(ctx context.Context, p *peer.Peer, group NeighborsGroup,
-	connectOpts ...ConnectPeerOption,
-) error {
-	m.log.Debugw("adding inbound neighbor", "peer", p.ID())
-	return m.addNeighbor(ctx, p, group, m.acceptPeer, connectOpts)
+	if err := m.peerDB.UpdatePeer(peer); err != nil {
+		return ierrors.Wrapf(err, "failed to update peer %s", peer.ID)
```

Comment:
Shouldn't we close the stream here as well?

Reply:
Yes we should, as we otherwise will leave a negotiated stream in the air without adding it to our list of known neighbors.